src/engine-actions/translate-swap.js

```diff
@@ -38,40 +50,141 @@ async function translateSwap (takerAddress, swapHash, amount, maxTime, finalCltv
+  // preimage for the provided swapHash, we can be confident that a call to
+  // sendPayment will result in creating only one downstream payment for this swap
+  try {
+    // we await so rejections are handled within this function
+    const paymentPreimage = await getPreimageIfPaymentExists(swapHash, this.client, this.logger)
+
+    return paymentPreimage
+  } catch (e) {
+    // If our call to LND returned a gRPC NOT FOUND error, we can be confident there
+    // isn't an existing payment for the provided swapHash, so we can create a new
+    // payment with a call to sendPayment
+    if (e.code !== grpc.status.NOT_FOUND) {
+      throw e
+    }
+
+    const { publicKey } = networkAddressFormatter.parse(takerAddress)
+
+    // Convert the maximum time into a time from now so we can convert it into
+    // blocks of relative time. We round down to a whole number of seconds to
+    // be conservative.
+    const timeFromNow = Math.floor((maxTime - new Date()) / 1000)
+
+    if (timeFromNow < 0) {
+      throw new PermanentSwapError(`Expected maxTime to be in the future. Seconds from now is: ${timeFromNow}`)
+    }
+
+    // Translate seconds into blocks, rounding down to ensure we can stay atomic.
+    // The LND api expects this to be an int32, which in javascript is a number.
+    const cltvLimit = parseInt(Big(timeFromNow).div(this.secondsPerBlock).round(3).toFixed(0), 10)
+
+    // Translate seconds into blocks, rounding up so that we're sure the end
+    // node will accept the payment. The LND api expects this to be an int32,
+    // so we convert to a javascript number.
+    const finalCltvDelta = parseInt(Big(finalCltvDeltaSecs).div(this.secondsPerBlock).round(0).toFixed(0), 10)
+
+    if (cltvLimit < finalCltvDelta) {
+      throw new PermanentSwapError(`Timelock for total swap is shorter than final hop of payment. ` +
+        `cltvLimt: ${cltvLimit}, finalCltvDelta: ${finalCltvDelta}`)
+    }
+
+    const {
+      paymentPreimage,
+      paymentError
+    } = await sendPayment({
+      paymentHash: swapHash,
+      destString: publicKey,
+      amt: amount,
+      finalCltvDelta,
+      cltvLimit,
+      feeLimit: DEFAULT_FEE_LIMIT
+    }, { client: this.client })
+
+    if (paymentError) {
+      this.logger.error(`Payment error while sending to route`, { paymentError, swapHash })
+      throw new PermanentSwapError(paymentError)
+    }
+
+    return paymentPreimage
   }
+}
+
+/**
+ * Retrieve the preimage for the given payment hash. If no payment is found
+ * for the hash, it will throw a gRPC NOT FOUND error. If the payment is
+ * in-flight, it will wait until the payment succeeds or is in a permanent
+ * failure state.
+ * @param {string} paymentHash - Base64 encoded payment hash
+ * @param {LndClient} client
+ * @param {Logger} logger
+ * @returns {Promise<string>} - Base64 encoded string of the preimage
+ */
+async function getPreimageIfPaymentExists (paymentHash, client, logger) {
```

Comment:
[nit] might be nice to have the second parameter as an object with properties `client`, and `logger`